bubblegum/program/src/lib.rs

```diff
@@ -665,6 +704,120 @@ fn process_creator_verification<'info>(
     )
 }
 
+fn process_collection_verification<'info>(
+    ctx: Context<'_, '_, '_, 'info, CollectionVerification<'info>>,
+    root: [u8; 32],
+    data_hash: [u8; 32],
+    creator_hash: [u8; 32],
+    nonce: u64,
+    index: u32,
+    mut message: MetadataArgs,
+    verify: bool,
+    sized: bool,
+) -> Result<()> {
+    let owner = ctx.accounts.owner.to_account_info();
+    let delegate = ctx.accounts.delegate.to_account_info();
+    let merkle_slab = ctx.accounts.merkle_slab.to_account_info();
+    let collection_metadata = &ctx.accounts.collection_metadata;
+    let collection_mint = ctx.accounts.collection_mint.to_account_info();
+    let edition_account = ctx.accounts.edition_account.to_account_info();
+    let collection_authority = ctx.accounts.collection_authority.to_account_info();
+
+    require!(
+        *collection_metadata.to_account_info().owner == ctx.accounts.token_metadata_program.key(),
+        BubblegumError::IncorrectOwner
+    );
+    require!(
+        *collection_mint.owner == spl_token::id(),
+        BubblegumError::IncorrectOwner
+    );
+    require!(
+        *edition_account.owner == ctx.accounts.token_metadata_program.key(),
+        BubblegumError::IncorrectOwner
+    );
+
+    // If the NFT has collection data, we set it to the correct value after doing some validation.
+    if let Some(collection) = &mut message.collection {
+        // Create a token-metadata Metadata struct and ONLY populate the collection field.
+        // This is being created only to interact with some token-metadata assert functions.
+        let mut mpl_metadata = mpl_token_metadata::state::Metadata::default();
+        mpl_metadata.collection = Some(collection.adapt());
+
+        // Collection verify assert from token-metadata program.
+        assert_collection_verify_is_valid(
+            &mpl_metadata,
```

Comment:
This works because the types all match:
-  mpl_metadata: I instantiate the NFT's metadata with the previous two lines of code (more on this at bottom).
- collection_metadata can be deserialized because `AccountSerialize` and `AccountDeserialize` were implemented for the struct `TokenMetadata(mpl_token_metadata::state::Metadata)` in metaplex_anchor.rs.
- collection_mint: This is just used for its key.
- edition_account: This is just account info, and it is actually deserialized inside of `assert_collection_verify_is_valid()`, but that token-metadata function knows how to do that.

Regarding the mpl_metadata that I instantiate, I could get away with not having to create that if I change the prototype of `assert_collection_verify_is_valid()` since it really only uses the collection information not any of the rest of the NFT's metadata.

See this idea for an example: https://github.com/metaplex-foundation/metaplex-program-library/pull/690

The reason I did not do it that way is because then I would have to wait for token-metadata to be published again before this contract picks up the updated prototype, and I am not sure of our timeline for token-metadata publishing.

Reply:
I my latest version I have to update token-metadata anyways, so the changes from https://github.com/metaplex-foundation/metaplex-program-library/pull/690 are now included in this PR.